src/main/java/uk/gov/pay/products/service/PaymentCreator.java

```diff
@@ -74,12 +78,38 @@ public Payment doCreate(String productExternalId) {
             paymentEntity.setExternalId(randomUuid());
             paymentEntity.setProductEntity(productEntity);
             paymentEntity.setStatus(PaymentStatus.CREATED);
-            paymentDao.persist(paymentEntity);
+            paymentEntity.setGatewayAccountId(productEntity.getGatewayAccountId());
+            paymentEntity.setReferenceNumber(randomUserFriendlyReference());
 
-            return paymentEntity;
+            AtomicInteger counter = new AtomicInteger(0);
+            return mergePaymentEntityWithReferenceNumberCheck(paymentEntity, counter);
         };
     }
 
+    private PaymentEntity mergePaymentEntityWithReferenceNumberCheck(PaymentEntity paymentEntity, AtomicInteger counter) {
```

Comment:
I don't see the need of AtomicInteger here given that its a method param thus making it stateless and threadsafe. A simple int would do.